reconcile.go

```diff
@@ -1,256 +1,216 @@
-	// required default setup as `PreparedConfig` under the response object.
-	// Warning: Diagnostics houses errors, the typical go err pattern isn't followed - must check `resp.Diagnostics.Err()`
-	prepConfigResp := provider.PrepareProviderConfig(providers.PrepareProviderConfigRequest{
-		Config: configFull,
-	})
-	if prepConfigResp.Diagnostics.Err() != nil {
-		log.Println(prepConfigResp.Diagnostics.Err().Error())
-		panic("Failed to prepare config")
-	}
-
-	// Lets set the values we need to set while we have the value map
-	configValueMap = prepConfigResp.PreparedConfig.AsValueMap()
-	configValueMap["client_id"] = cty.StringVal(os.Getenv("ARM_CLIENT_ID"))
-	configValueMap["client_secret"] = cty.StringVal(os.Getenv("ARM_CLIENT_SECRET"))
-	configValueMap["tenant_id"] = cty.StringVal(os.Getenv("ARM_TENANT_ID"))
-	configValueMap["subscription_id"] = cty.StringVal(os.Getenv("ARM_SUBSCRIPTION_ID"))
-
-	// Now we have a prepared config we can configure the provider.
-	// Warning (again): Diagnostics houses errors, the typical go err pattern isn't followed - must check `resp.Diagnostics.Err()`
-	configureProviderResp := provider.Configure(providers.ConfigureRequest{
-		Config: cty.ObjectVal(configValueMap),
-	})
-	if configureProviderResp.Diagnostics.Err() != nil {
-		log.Println(configureProviderResp.Diagnostics.Err().Error())
-		panic("Failed to configure provider")
-	}
-}
-
 // TerraformReconciler is a reconciler that processes CRD changes uses the configured Terraform provider
 type TerraformReconciler struct {
 	provider *plugin.GRPCProvider
+	client   client.Client
 }
 
-func NewTerraformReconciler(provider *plugin.GRPCProvider) *TerraformReconciler {
+func NewTerraformReconciler(provider *plugin.GRPCProvider, client client.Client) *TerraformReconciler {
 	return &TerraformReconciler{
 		provider: provider,
+		client:   client,
 	}
 }
 
-func (r *TerraformReconciler) Reconcile(crd *unstructured.Unstructured) error {
+func (r *TerraformReconciler) Reconcile(ctx context.Context, log logr.Logger, crd *unstructured.Unstructured) (*ctrl.Result, error) {
 
+	log.Info("Reconcile starting")
 	// Get the kinds terraform schema
 	kind := crd.GetKind()
 	resourceName := "azurerm_" + strings.Replace(kind, "-", "_", -1)
 	schema := r.provider.GetSchema().ResourceTypes[resourceName]
 
 	var configValue *cty.Value
 	var deleting bool
-	if deleting = isDeleting(crd); deleting {
+	if deleting = r.isDeleting(crd); deleting {
 		// Deleting, so set a NullVal for the config
 		v := cty.NullVal(schema.Block.ImpliedType())
 		configValue = &v
 	} else {
-		ensureFinalizer(crd)
+		if err := r.ensureFinalizer(ctx, log, crd); err != nil {
+			return reconcileLogError(log, fmt.Errorf("Error adding finalizer: %s", err))
+		}
 
 		// Get the spec from the CRD
-		spec := crd.Object["spec"].(map[string]interface{})
+		spec, gotSpec, err := unstructured.NestedMap(crd.Object, "spec")
```

Comment:
oh yeah this is nicer with the `NestedMap` helpers

Reply:
Yeah - they were a good call. Thanks for pointing them out!